sql/src/main/java/io/crate/metadata/sys/SysSnapshotsTableInfo.java

```diff
@@ -21,73 +21,43 @@
-    private static final RowGranularity GRANULARITY = RowGranularity.DOC;
 
-    public static class Columns {
-        public static final ColumnIdent NAME = new ColumnIdent("name");
+    static SystemTable<SysSnapshot> create() {
+        return SystemTable.<SysSnapshot>builder()
+            .add("name", STRING, SysSnapshot::name)
+            .add("repository", STRING, SysSnapshot::repository)
+            .add("concrete_indices", STRING_ARRAY, SysSnapshot::concreteIndices)
+            .add("started", TIMESTAMPZ, SysSnapshot::started)
+            .add("finished", TIMESTAMPZ, SysSnapshot::finished)
+            .add("version", STRING, SysSnapshot::version)
+            .add("state", STRING, SysSnapshot::state)
+            .add("failures", STRING_ARRAY, SysSnapshot::failures)
+            .setPrimaryKeys(new ColumnIdent("name"), new ColumnIdent("repository"))
+            .withRouting(SysSnapshotsTableInfo::getRouting)
+            .build(IDENT);
     }
 
-    static Map<ColumnIdent, RowCollectExpressionFactory<SysSnapshot>> expressions() {
-        return columnRegistrar().expressions();
-    }
-
-    private static ColumnRegistrar<SysSnapshot> columnRegistrar() {
-        return new ColumnRegistrar<SysSnapshot>(IDENT, GRANULARITY)
-          .register("name", STRING, () -> forFunction(SysSnapshot::name))
-          .register("repository", STRING, () -> forFunction(SysSnapshot::repository))
-          .register("concrete_indices", STRING_ARRAY, () -> forFunction(SysSnapshot::concreteIndices))
-          .register("started", TIMESTAMPZ, () -> forFunction(SysSnapshot::started))
-          .register("finished", TIMESTAMPZ, () -> forFunction(SysSnapshot::finished))
-          .register("version", STRING, () -> forFunction(SysSnapshot::version))
-          .register("state", STRING, () -> forFunction(SysSnapshot::state))
-          .register("failures", STRING_ARRAY, () -> forFunction(SysSnapshot::failures));
-    }
-
-    SysSnapshotsTableInfo() {
-        super(IDENT, columnRegistrar(), "name","repository");
-    }
-
-    @Override
-    public RowGranularity rowGranularity() {
-        return GRANULARITY;
-    }
-
-    @Override
-    public RelationName ident() {
-        return IDENT;
-    }
-
-    @Override
-    public Routing getRouting(ClusterState clusterState,
-                              RoutingProvider routingProvider,
-                              WhereClause whereClause,
-                              RoutingProvider.ShardSelection shardSelection,
-                              SessionContext sessionContext) {
+    private static Routing getRouting(DiscoveryNodes nodes) {
         // route to random master or data node,
         // because RepositoriesService (and so snapshots info) is only available there
-        return routingProvider.forRandomMasterOrDataNode(IDENT, clusterState.getNodes());
+        return new RoutingProvider(new Random().nextInt(), List.of()).forRandomMasterOrDataNode(IDENT, nodes);
```

Comment:
I think we should extend the `SystemTable` `getRouting` method and add the `routingProvider` argument.

The `routingProvider` is initialized **once** per plan/query with the same seed, so that we have a stable routing result within one query. 

May not matter much for system tables, but I think otherwise the difference in behavior could be confusing.

Reply:
Sounds good, i will do a separate pr for this.

Reply:
I think it should be part of this PR, as otherwise this PR introduces a behavior change and is no longer just a internal refactoring.

Reply:
👍